### ISSUE TITLE
Subsequent calls to requestPaymentMethod do not result in new nonces

### ISSUE THREAD

**Issue by idontusenumbers** [braintree/braintree-web-drop-in#363]
### General information

* SDK/Library version: 1.9.4
* Environment: Sandbox
* Browser and OS Chrome 64.0.3282.167 (64-bit) on MacOS 10.13.2 -->

### Issue description

Subsequent calls to instance.requestPaymentMethod() do not result in new nonces unless a different payment method is selected. This can cause 'nonce can only be used once' errors when submitting.

**Comment by crookedneighbor** [comment]
Can you explain your use case a little more? Why do you need a second nonce?

**Comment by idontusenumbers** [comment]
The flow right now with the dropin is already a little awkward. There's both a 'submit payment method' step then a 'submit purchase' step. There's no way to know if the user has entered a new card (when it would make sense to first confirm the payment method then submit the purchase), or if the user has selected (or defaulted) to a saved payment method and just want to move forward with the purchase. My solution right now is to just auto submit the purchase and not give them an option to select another payment method.

Anyway, that doesnt totally explain the issue reported here; I mostly encounter it while testing. it leaves me with a 'this api is buggy/inconsistent' feeling that, although maybe not a valid use case, should be shored up for consistency. I feel all calls to requestPaymentMethod should result in a rejection OR a valid nonce. Having it be 'sometimes results in a valid nonce, sometimes a duplicate, who knows, there's no way to detect' has bad code smell to me.

**Comment by crookedneighbor** [comment]
The basic issue is that there are 2 kinds of payment methods that Drop-in supports. Those that require the user to fill out a form from within Drop-in (credit cards), and those that require the user to go through a flow outside of Drop-in (PayPal, Apple Pay, etc). The first type tokenizes the data by calling `requestPaymentMethod`, while the second just returns the nonce that has already been tokenized by going through the external flow.

You do not need to buttons to make this process smooth. You can use the events API. https://braintree.github.io/braintree-web-drop-in/docs/current/Dropin.html#on

Take a look specifically at the `Automatically submit nonce to server as soon as it becomes available` example.

This will allow you to send the nonce as soon as it becomes available (such as in the case of PayPal or Apple Pay) or let the customer submit it when the form is filled out (in the case of the card form). Once you've sent the nonce to the server and transacted with it, you'll want to navigate to a new page or hide the Drop-in.

I'm going to close this now, because I think this will help improve your set up, but feel free to comment here if you have additional questions about it.

**Comment by idontusenumbers** [comment]
Another use case:
User enters a credit card number for a card that has been deactivated or would otherwise be rejected by the gateway.

User clicks purchase, app hits braintree, braintree rejects, app shows user message that the card was declined.

User wants to try again to to make sure it wasn't a fluke. User sees the previously used card selected and clicks purchase again. App shows exception from braintree saying the Nonce can only be used once, which is confusing to the user.

User thinks 'maybe I need to enter it again', enters it again, and sees no new card added to the list but the previously entered one is still selected. This is confusing.

**Comment by crookedneighbor** [comment]
For that use case, I'd print an error message to your customer and use [`clearSelectedPaymentMethod`](https://braintree.github.io/braintree-web-drop-in/docs/current/Dropin.html#clearSelectedPaymentMethod).

I've opened a PR to include an option in the card configuration to persist card details after they have been tokenized to make it easier for customers to review their payment details in the case of a transaction sale error: https://github.com/braintree/braintree-web-drop-in/pull/373

**Comment by idontusenumbers** [comment]
clearSelectedPaymentMethod() removes the card and resets the form which means the user has to enter the information again. This saves them from the latter confusion but still doesn't allow for the the simple 'just try again' course of action which I think most people do when something like this fails; for example when a card is declined at mcdonalds.

**Comment by saoudrizwan** [comment]
Is there any progress on this, or a suggested workaround?
Let's say that a customer has 1 method of payment for everything on my site. And I use this drop-in on a settings page that lets him change his default method of payment. Each time he changes the payment method (which I detect with the `paymentMethodRequestable` event) I submit the nonce generated from `.requestPaymentMethod()` to my server. However, if the user deselects a payment method and then selects it again, a duplicate nonce is generated, and my server complains about nonce being used more than once. 
A fix for this would be to _always_ generate a unique nonce when calling `.requestPaymentMethod()`, as per the definition of nonce: number used only **once**.

**Comment by saoudrizwan** [comment]
I figured out a workaround for anyone wondering. 

On the client's side, I maintain a cache object `[usedNonce: tokenReturnedFromServer]`, where `tokenReturnedFromServer` is the resulting [`token`](https://developers.braintreepayments.com/reference/request/payment-method/create/node#token) from using the nonce in a Payment Method action. 

Now whenever I want to make a request that uses a nonce, I check this dictionary first to see if I already used this nonce, if I did I pass the associated token instead (else I pass the nonce like normal). On the server's side, you have to handle whether the client gives you a nonce or a token, and you'll be sure that you'll get a token instead of a nonce that's been used before.

Also, this cache of nonce to tokens can be purged as soon as the drop-in is torn down, since at that point, `.requestPaymentMethod()` returns fresh new nonces.

**Comment by crookedneighbor** [comment]
Another workaround is simply to [teardown](https://braintree.github.io/braintree-web-drop-in/docs/current/Dropin.html#teardown) and then re-set up Drop-in. Once that is done, any vaulted payment methods will have new nonces associated with them. Any payment methods that were not vaulted will be removed.

The long term solution for this will be to have a method to reset Drop-in so it can simply refresh the vaulted payment method nonces with unconsumed ones. Under the hood, this would just be fetching the array of saved payment methods for the customer id that was passed in, and replacing the existing payment methods with the fresh ones.

Going to re-open this issue in the meantime.

**Comment by saoudrizwan** [comment]
Just found out that `.clearSelectedPaymentMethod()` will generate new nonces on any subsequent `.requestPaymentMethod()` calls.

**Comment by crookedneighbor** [comment]
Ahh, yes. I forgot we added that. That's exactly what that was intended for. Going to close this. Thanks for bringing it up.

**Comment by joaohaas** [comment]
Sorry for necroing, but doesn't the issue mentioned by @/idontusenumbers about `clearSelectedPaymentMethod` still stand? That is, the expectation is that if a payment fails, ideally you'd show an error message (e.g. 'Insufficient Funds'), the user would solve whatever is happening, and would retry.

Calling `clearSelectedPaymentMethod` unselects the current payment method. I think either a separate method that only clears the nonce or an option to pass to `requestPaymentMethod` to ignore the cached value would greatly help this kind of error handling flow.
